lib/src/back_end/solution.dart

```diff
@@ -4,9 +4,11 @@
 import '../piece/piece.dart';
 import 'code_writer.dart';
 
-/// A possibly incomplete set of selected states for a set of pieces being
-/// solved.
-class PieceStateSet {
+/// A single possible line splitting solution.
```

Comment:
[nit] What is the word "single" referencing here? Since the field is `final` in the code writer, I expect it's mutable and the splits can change over the lifetime of this object - is that correct?

Reply:
It's single in that Solver explores a tree of Solutions. Each Solution has a fixed set of States that determines how it formats the code. (This will change *sort of* in an upcoming optimization.)

The splits don't change over the lifetime of any given Solution. But the Solver produces new Solutions derived from earlier ones and the new ones have different splits than their parents.

It's a little hard to explain in a code review comment, but at a high level, the Solver is doing [combinatorial search](https://en.wikipedia.org/wiki/State_space_search). It's trying to reach a goal state where every Piece has a chosen State. It starts the search from an initial Solution where no Piece is bound to a state. It reaches new Solutions by taking an existing Solution picking an unbound Piece from it, and making a set of new Solutions where each new Solution inherits the bound States of its parent and then also binds the chosen Piece to each of its possible states.